### ISSUE TITLE
What's a reasonable expectation of angular performance?

### ISSUE THREAD

**Issue by fxck** [angular/angular#38184]
So my app is slow. That slow part consists of iteration over medium complex components with another bunch of nested components inside
![image](https://i.imgur.com/aoyc1er.png)

Trying to render more than one of these started to add up a LOT of scripting time, up to the point where trying to render 10 of them caused 30 second freeze. 

The first and biggest culprit I found was @angular/flex-layout, which, as you know, if you managed to click [deep enough](https://github.com/angular/flex-layout/wiki/Performance-Considerations) inside the wiki can not handle more than a hundred of grids (since I used it inside like 10 of the nested components inside it easily added up to that critical number).

But even still, with flex-layout completely gone, trying to render 100 of these can take around 12 seconds, so I thought there must be sometimes else wrong and tried to establish a baseline on a completely fresh project. That can be found here https://github.com/fxck/ngperf.

### Baseline
Iterating over 200 items, printing simple text.

![image](https://user-images.githubusercontent.com/1303561/88206816-2c8ad480-cc4f-11ea-985b-44441afba764.png)

That takes around **100ms** from click on the button to the content being visible.

![image](https://user-images.githubusercontent.com/1303561/88206895-4cba9380-cc4f-11ea-927e-4a40fa48ae3d.png)

### Baseline for complex tests
Iterating over 200 components with mat-card and ngClass inside
![image](https://user-images.githubusercontent.com/1303561/88207043-7d023200-cc4f-11ea-81ca-0da3bebffe8d.png)

![image](https://user-images.githubusercontent.com/1303561/88207075-8ab7b780-cc4f-11ea-9e99-923fac814cc1.png)

That takes around **200ms** from click on the button to the content being visible.

![image](https://user-images.githubusercontent.com/1303561/88207131-a327d200-cc4f-11ea-841b-d7ce264a261a.png)

### Test 1
Adding 20 empty divs to complex baseline

![image](https://user-images.githubusercontent.com/1303561/88207220-c6eb1800-cc4f-11ea-8442-7a25a65a61ca.png)

That takes around **250ms** from click on the button to the content being visible, most of the extra time taken by rendering, which makes sense.

![image](https://user-images.githubusercontent.com/1303561/88207323-e3875000-cc4f-11ea-953b-4fc7222a34d0.png)

### Test 2
Adding 20 divs with ngClass and ngStyle on them

![image](https://user-images.githubusercontent.com/1303561/88207431-129dc180-cc50-11ea-849e-a8d5c2a441ba.png)

That takes around **400ms** from click on the button to the content being visible, extra time coming from scripting.. given that's total of like 4000 new ngClasses and ngStyles, 100ms extra time spent on scripting is fair I guess
![image](https://user-images.githubusercontent.com/1303561/88207510-319c5380-cc50-11ea-9bbd-7acbea73360e.png)
 
### Test 3 
Adding 20 components with a single div with ngClass and ngStyle inside them

![image](https://user-images.githubusercontent.com/1303561/88207807-9b1c6200-cc50-11ea-856c-b518ffc5085a.png)
![image](https://user-images.githubusercontent.com/1303561/88207841-a53e6080-cc50-11ea-8b99-64109d383bd7.png)

Now this takes around **900ms** from click on the button to the content being visible, 200 extra milliseconds coming rendering and another 200ms from scripting.

![image](https://user-images.githubusercontent.com/1303561/88207954-cdc65a80-cc50-11ea-8bfe-57661fa80f75.png)

The third one confuses me the most, why would it take double the amount of time between test 2 and test 3? That makes me question everything about splitting my app into many reusable component, if there's a huge performance hit like this.

Coming back around to my actual component (https://i.imgur.com/aoyc1er.png), it has quite a few nested components, most of them have multiple states, content projection, styles, classes, so seeing how much mess can an empty component create, how many can I possibly render at once before having to resort to infinite scroll strategies?

(both repro and my app uses v10)
![image](https://user-images.githubusercontent.com/1303561/88208529-a623c200-cc51-11ea-93c7-125d29d5088b.png)

---

btw this is what rendering 20 of those components looks like WITHOUT flex-layout

![image](https://user-images.githubusercontent.com/1303561/88210283-3b27ba80-cc54-11ea-8329-723472062dff.png)

and WITH flex
![image](https://user-images.githubusercontent.com/1303561/88210608-bf7a3d80-cc54-11ea-9524-9d8397f36fd4.png)

20 seconds of scripting time difference!

**Comment by fxck** [comment]
@mgechev told me on Twitter it's not really surprising

![image](https://user-images.githubusercontent.com/1303561/88232521-4ee41880-cc76-11ea-9556-a514507cc1e5.png)

So I thought I'd give it a try and rewrite test 3 in react, haven't really touched react before, so I'm not sure if it's done right https://github.com/fxck/reactperf/blob/master/apps/perf/src/app/app.tsx 

either way, it's about 100ms compared to 900ms in angular.

![image](https://user-images.githubusercontent.com/1303561/88232698-9d91b280-cc76-11ea-9c02-b33170e925cb.png)

**Comment by mgechev** [comment]
Keep in mind that this is one use case and a single benchmark. There are hundreds of others (content projection, different bindings, directives vs components, etc).

At the same time, this brings much more than rendering. Dependency injection and view encapsulation are included out of the box.

As I mentioned on Twitter @pkozlowski-opensource will be able to share more insights. He has been digging a lot into initial rendering performance.

On Twitter I was primarily curious what is the expected resolution of this issue? Seems more appropriate for a discussion on gitter, Twitter or another medium.

**Comment by hiepxanh** [comment]
sorry for the irrelevant question but,
@fxck  the first image, which platform are you deploy?

**Comment by fxck** [comment]
> On Twitter I was primarily curious what is the expected resolution of this issue? Seems more appropriate for a discussion on gitter, Twitter or another medium.

I've been discussing it on angular discord for last week, everyone was surprised, none initially believed it should be a problem to render 100 of my components, let alone empty components. I guess the resolution would be confirmation that there is nothing odd going on and that increase between 2 and 3 is to be expected.

@hiepxanh what do you mean? What we use for deploy? The thing itself, it's a platform that builds, deploys and runs your code as well as databases and storages and then takes care of auto scaling and fine tuning. It's barely an alpha though, should be ready sometime at the beginning of the next year.

**Comment by fxck** [comment]
btw @mgechev 

this is how long it takes to render empty `<div></div>` inside `<app-test></app-test>` component with encapsulation turned off just in case as well.

![image](https://user-images.githubusercontent.com/1303561/88261826-52ec5680-ccc7-11ea-8186-f618111200bf.png)

And this is how long it takes to render 
```
  <div
    [ngClass]="{
      'foo': true
    }"
    [ngStyle]="{
      'width.px': 100,
      'width.px': 100,
      'background': '#ccc'
    }">
  </div>
```

directly instead..

![image](https://user-images.githubusercontent.com/1303561/88261916-7911f680-ccc7-11ea-8e50-b88ace6d18d1.png)


Are you 100% sure that doesn't look fishy? I can't help but feel this implies you should create one huge component instead composing your app of multiple small ones. Which is something I never heard before.

**Comment by pkozlowski-opensource** [comment]
@fxck I'm starting to look into your scenarios (for now beginning with comparing baseline with the iterating over `<app-simple>`) and I'm getting substantially different numbers from the ones you've posted. I will obviously keep digging and will share my findings here, but just to verify the obvious (I'm sorry, I know that those are basics but I've seen so many skewed perf measurements in the past that I prefer to verify):
* are you running your application in the prod mode when taking measurements (`ng build --prod`)? The dev mode has significant overhead, so I wouldn't expect numbers to be any good;
* are you taking measurements in the incognito mode with all the extensions switched off? Many browser extensions can interact with the page and skew the results significantly.

For the record I'm seeing ~50% difference (~9ms -> ~14ms scripting time) moving from baseline -> simple on the initial render (going from an empty list to a rendered one). Off the bat I would say it is in-line with my expectations as:
* there is 2x amount of DOM nodes (div + app-simple so we are not really comparing the same generated pages);
* components _do_ have cost (I will expand on it later).

In any case - as of right now I'm seeing significant discrepancy between your and my numbers so we need to clarify this first.

**Comment by pkozlowski-opensource** [comment]
OK, so there are many scenarios being discussed here and many measurements so let's start tackling them one by one. Let's start with the baseline (200 `<div>` elements ) -> simple (200 `<app-simple>` instances).

First of all, as indicated previously, I don't think we compare apples to apples here as the DOM created is very different in both scenarios:
* 200 x `<div class="item">raw</div>` in the baseline
* 200 x `<div class="item"><app-simple>simple</app-simple></div>` - notice that we are creating 2x amount of DOM nodes here.

Let's put the difference aside for the second and have a look at the numbers I'm getting (scripting for first creation in the incognito mode, production build):
* baseline: ~10ms 
* simple: ~14ms (~45% overhead).

See pictures below:

![Screen Shot 2020-07-23 at 14 13 58](https://user-images.githubusercontent.com/973550/88285079-c8b7e880-ccee-11ea-9a43-8aa64ca0719b.png)

![Screen Shot 2020-07-23 at 14 15 52](https://user-images.githubusercontent.com/973550/88285280-1c2a3680-ccef-11ea-9156-38fb60cc65a4.png)


Again, the DOM produced is not the same and the DOM nodes creation + connecting together takes time. If we change the app to generate the same amount of nodes like so:

```html
<app-simple class="item" *ngFor="let item of data; trackBy: trackByIndex;"></app-simple>
```

the numbers get much closer:
* baseline: ~10ms (as before)
* simple: ~11.5ms (~18% overhead).

See picture below:

![Screen Shot 2020-07-23 at 14 19 09](https://user-images.githubusercontent.com/973550/88285753-ed609000-ccef-11ea-9f44-a4dccf514c17.png)

Obviously the numbers fluctuate a lot from one run to another (this is another story worth several blog posts) but TBH I'm surprised that we see < 20% overhead here given all the work we need to do to instantiate a component  (it is much work work than just creating an instance of a class - we are allocating internal data structures, resolve DI injection, take care of CSS encapsulation etc.). 

Components are definitively not "free" - especially in the case you are demonstrating here (where each DOM node is replaced by a component). 

I can share profile recordings with the above numbers as needed.

For now I'm afraid that I ca't reproduce the same numbers as reported (at least not for the case discussed in this comment). We will need to understand why we are getting different numbers and as noted before I would expect either a dev mode or browser extensions - to be investigated. @fxck could you please check the settings when capturing measurements?

I will keep looking at the other scenarios as well but yeh, we need to gain more trust in the numbers shared initially.

**Comment by pkozlowski-opensource** [comment]
@fxck I've also had a look at another scenario of yours comparing the cost of components with ngClass / ngStyle (scenario desired here https://github.com/angular/angular/issues/38184#issuecomment-662862655).

Sadly, I can't replicate the numbers you are getting :-/ The numbers I get a more like:
* instances of the `<app-test>` with a div inside (no ngStyle / ngClass involved): ~130ms
* divs with ngStyle / ngClass: ~185ms
* instances of the `<app-test>` with a div inside _and_ ngStyle / ngClass: ~320ms

which kinds of adds up. Again, components are far from being "free" in the initial creation but I'm getting slightly different picture from my measurements. On a separate note `ngClass` / `ngStyle` _are_ relatively expensive so all those scenarios are exercising the framework in  rather particular conditions. 

I did another experiment where I've replaced ngClass / ngStyle with a class binding (there was only one dynamic part):

```html
<div [class.foo]="true" style="
    width: 100px;
    height: 100px;
    background: #ccc;">
</div>
```

and then numbers dropped from ~320ms down to ~230ms which again shows the cost of directives (and ngClass / ngStyle are relatively expensive as they need to diff css classes / styles). I'm not sure if you can apply the same changes in your application but if so limiting usage of static CSS classes / styles in ngClass / ngStyle can have positive effect on perf.

The bottom line: all the scenarios discussed here are heavy on the DOM (20k+ nodes) and each of them having 2 quite expensive directives. Plus there is some costs to components.

I'm not sure if I can say more at this point - what is worrying for me is that we are getting very different numbers from our measurements so it is hard to have trust in them. But at least I hope that some of my input might be useful in clarifying hints / giving you some ideas on what to shuffle in your application.

Happy to continue the discussion but we need to gain more trust in the numbers discussed here so I would like to ask you to check the dev mode and browser extensions.

**Comment by fxck** [comment]
@pkozlowski-opensource thank you for looking into this, really really appreciate it, I'm off on a holiday for the rest of the week and the next one, I'll get back to you after that.. `prod` could definitely be a factor, I did test everything multiple times with different settings and mode, not sure what I screenshot'd in the end. I 100% remember that there was literally no difference between prod and JIT (funnily enough JIT produced faster results) in my actual app, but that was using flex at the time, so that could simply be because flex was causing such major lags that those small difference were lost in it.

**Comment by pkozlowski-opensource** [comment]
@fxck sure, no hurry - if you can re-measure when you are back we could get better understanding of what is going on. I'm definitively more than interested in digging into the topic, but also want to set expectations: lots of nodes with lots of directives will have cost. 

> I 100% remember that there was next to no difference between prod and JIT (funnily enough JIT produced faster results) in my actual app, but that was using flex at the time, so that could simply be because flex was causing such major lags that those small difference were lost in it

Specifically I would like to reproduce this one, as it sound very odd. 
Anyway, enjoy your time off, let's get together when you are back!

**Comment by hiepxanh** [comment]
@fxck I see your platform is really beautiful. Feeling like first impression of mystery girl. Can you share me some link and information? I would like to subscribe to that platform. (because it beautiful 😄 )

BTW, I think you should make a loom recording, maybe that can help?

**Comment by fxck** [comment]
@hiepxanh thank you, we'll have a teaser site out sometime next week, with the option to subscribe for news, I'll send you link then.

**Comment by petebacondarwin** [comment]
@fxck - are you able to provide further information about this performance issue?

**Comment by fxck** [comment]
@petebacondarwin I had to move forward and the code is quite different now, I might be able to create a demo using some old version, but not in foreseeable future, sorry :/ . But I'm sure the most impactful was flex-layout, I moved away from it completely, along with simply showing only few of the most complex components at the same time pretty much "fixed" it.

**Comment by petebacondarwin** [comment]
Thanks for the update @fxck - I think in that case we can close this issue.
